### PR TITLE
fix(agent): let invoked child agents serve their own first-party tools

### PR DESCRIPTION
**Pull request by kwakayama**
## Problem

The secondary finding from veryfront/veryfront-issue-inbox#1327: in local runtimes (`veryfront schedule run`, local dev), every `invoke_agent` child that declares first-party file tools fails with `Unknown tool references: create_file, get_file` — even though the child's own agent definition authorizes them and the host bootstrap credential is present. This makes multi-agent templates undemoable locally.

## Root cause

Nested tool execution wraps children in the invoking runtime's remote tool sources, constrained to the INVOKER's allowlist (`constrainRuntimeRemoteToolSources` → `runWithRuntimeRemoteToolSources`). In `getRuntimeRemoteToolSources`, the inherited source's id (`veryfront-platform-mcp`) then suppressed creation of the child's own bootstrap veryfront-api source, so a child of an orchestrator that only allows `list_files` can never reach `create_file`/`get_file`. Verified with instrumentation: child saw `implicit=["create_file","get_file"] inherited=1`, then resolved zero file tools.

## Fix

A child agent whose config names first-party tools builds its own bootstrap-identity source for those tools — exactly what it does when no parent context is active. Two boundaries preserved:
- A host that injected sources explicitly (`__vfRemoteToolSources`, the hosted/control-plane wiring) still owns the boundary — hosted behavior unchanged.
- When no bootstrap identity exists, the child keeps inherited-only behavior (unchanged).
Inherited sources whose id the child now self-serves are dropped to avoid duplicate source ids.

## Red/green

New test `serves a child agent's own named tools past a constrained inherited source`: red on main (resolves `[]`), green with the patch (`["create_file","get_file"]`). All 29 tests in the file pass; `deno check`, `deno lint`, `deno fmt --check` clean.

## Review rounds

Two further boundary defects were found in review and fixed on this branch:

- A ceiling stamped into `__vfAllowedRemoteTools` is an authorization cap, never an agent declaration: it is set only by a run (forwarded grants intersected with a runtime tool allowlist, the hosted request handler, the hosted chat runtime, fork authorization, AG-UI retirement limits). Propagating bootstrap provenance onto that wrapper let a child classify the cap as replaceable and rebuild the API source from the process bootstrap identity, reaching a tool the run's grant excluded. `constrainRuntimeRemoteToolSources` no longer propagates provenance, so the cap reads as owned. Credential-owner binding and the retained-alias wrapper still do, which is what the local scenario needs: that path stamps no ceiling at all.
- The injected-source ownership classification now skips non-own indexes, matching every helper in `private-array.ts`. A sparse ambient array could otherwise read a source off a patched `Array.prototype` and mark a real bootstrap-owned id as host-owned, stripping the child's remote tools when it cannot re-derive.

Both have regression tests verified red before the fix and green after. 34 tests in the touched file pass, plus 577 in `src/agent/runtime/`, 176 in `tests/integration/agent/`, 25 in `src/agent/streaming/`, 13 in `src/internal-agents/`. `deno check src/agent/index.ts`, `deno lint src/agent/runtime/`, `deno fmt --check` and the repo's test-shape lints are clean.

End-to-end verification in progress: the agentic-inbox-processing-outlook demo pipeline (orchestrator + 4 specialists, 6 emails) running locally against this branch.

https://claude.ai/code/session_019B9SYdGjpN6sQ4424NLv91

## Summary by CodeRabbit

- **Bug Fixes**
  - Child agents can access their own named tools even when inherited tool sources are restricted.
  - Host-provided tool sources remain authoritative, preventing unexpected replacement or re-derivation.
  - Nested runtimes now preserve correct tool-source ownership and access behavior across policy boundaries.
  - Conflicting inherited and host-provided sources are resolved consistently, keeping host tools available while excluding redundant sources.
  - Tool access limits remain enforced for invoked child agents, and inherited object properties no longer affect tool-source selection.

### CODE REVIEW

**Review by kojiwakayama — APPROVED** [review]

**Review by kwakayama — COMMENTED** [review]
## Review: 84/100 — solid, well-scoped fix; a couple of loose ends before merge

Traced the logic in `getRuntimeRemoteToolSources` end-to-end against the diff and the existing test suite. The core fix is correct and the security-sensitive boundary is handled carefully.

**Strengths**
- The root cause analysis is accurate: `selectedInjectedSources`/`configuredSources` previously suppressed a child's own bootstrap `veryfront-api` source purely because an *inherited* source shared its id, with no way for the child's own declared tools to win. The new `hostOwnedSourceBoundary` / `selfServedFirstPartyIds` split fixes exactly that, only when there's no explicit host injection (`__vfRemoteToolSources`) — so hosted/multi-tenant behavior (the actual privilege boundary) is untouched. I verified this against the existing `__vfRemoteToolSources`-based tests (e.g. "reuses an injected Veryfront API source", "enforces policy on injected Veryfront API source") and they still hold under the new code path.
- Good fallback behavior: when no bootstrap identity is available (`createVeryfrontApiMcpServerToolSource` returns `undefined`), `selfServedFirstPartyIds` is never populated for that id, so the old inherited/policy-wrapped source is correctly retained instead of silently dropped. This is exercised (indirectly) by the pre-existing "constrains inherited sources to implicit named tools" test.
- New test directly targets the fixed scenario (constrained inherited source + child's own bootstrap identity) with a clear red/green story per the PR description.
- Clear in-code comment explaining *why* the boundary split exists, not just what it does — will save the next reader from re-deriving this.

**Concerns**
1. **Unrelated lockfile churn**: `deno.lock` gains `"npm:zod@*": "4.3.6"` alongside the existing `"npm:zod@4.3.6"` entry, which is unrelated to this fix (nothing in the diff introduces a wildcard zod import). This looks like incidental drift from the local toolchain run. Worth regenerating/pruning before merge so it doesn't mask a real dependency change in a future diff.
2. **No test for the "child declares only some of its needed tools" edge case**: once a first-party source id is self-served, the corresponding *inherited* source for that id is dropped wholesale (`policyWrappedInjectedSources` filters it out). If a child relies on an ambient/inherited tool it does *not* itself declare in `tools` (e.g. only declares `create_file`/`get_file` but was implicitly relying on inherited `list_files`), that tool now silently disappears instead of being merged. This may be intended (tools should be declared on the child), but it's a behavior change worth a test and a callout in the PR description rather than leaving it implicit.
3. PR description says "End-to-end verification in progress" for the actual multi-agent demo pipeline — that's the most convincing validation of the real-world bug this is fixing, so it'd be good to confirm it completed successfully before merging rather than relying solely on the unit test.
4. Minor: `mergeable_state` is currently `blocked` and CI is `pending` on the head SHA — nothing wrong with the code, just flagging that this isn't merge-ready yet independent of review.

Nothing here blocks approval in principle — items 1–2 are cheap to address, item 3 is about closing the loop on the author's own stated validation plan.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 1]
```diff
@@ -376,22 +376,23 @@ export function getRuntimeRemoteToolSources(
...
-      (implicitToolNames.length > 0 ? { allow: implicitToolNames } : undefined);
-    return createMcpToolPolicySource(source, policy);
-  });
+  // Ambient inheritance carries the invoking runtime's sources, already
+  // constrained to the INVOKER's tool allowlist. A child agent whose own
+  // config names first-party tools must not be capped by that boundary: it
+  // builds its own bootstrap-identity source for its declared tools, exactly
+  // as it would when no parent context is active. Only a host that injected
+  // sources explicitly (__vfRemoteToolSources) still owns the boundary.
+  const hostOwnedSourceBoundary = configuredInjectedSources !== undefined;
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts]: Addressed in 9850e2c3e: ownership is now decided by source provenance, not child config. Sources the runtime builds from the host bootstrap identity are tracked by object identity (WeakSet) through the constrain and credential-binding wrappers; a child re-derives its own source only when every ambient same-id source carries that provenance. A host-injected source — explicit or ambient — keeps exclusive ownership of its id, and with an ambient same-id source present a missing bootstrap identity falls back to the inherited source instead of throwing. Regression test added: a host-injected ambient source stays authoritative and the bootstrap factory is provably never invoked.

**Comment by kwakayama** [comment]
@codex review

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 2]
```diff
@@ -145,6 +145,35 @@ function createMcpToolPolicySource(
...
+ */
+const bootstrapIdentityToolSources = new WeakSet<RemoteToolSource>();
+
+export function markBootstrapIdentityRemoteToolSource(source: RemoteToolSource): RemoteToolSource {
+  bootstrapIdentityToolSources.add(source);
+  return source;
+}
+
+export function isBootstrapIdentityRemoteToolSource(source: RemoteToolSource): boolean {
+  return bootstrapIdentityToolSources.has(source);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts]: Fixed in effdb3c2c: provenance membership now uses the repo's private weak store (captured WeakMap intrinsics via Reflect.apply at module load) instead of a raw WeakSet, so a patched WeakSet.prototype can neither launder a host-injected source into bootstrap ownership nor break child tool resolution.

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 3]
```diff
@@ -376,31 +411,47 @@ export function getRuntimeRemoteToolSources(
...
+      const sameIdSources = filterPrivateArray(
+        injectedSources,
+        (source) => source.id === sourceId,
+      );
+      const hostOwnsSourceId = configuredInjectedSources !== undefined
+        ? sameIdSources.length > 0
+        : somePrivateArray(
+          sameIdSources,
+          (source) => !isBootstrapIdentityRemoteToolSource(source),
+        );
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts:449]: Fixed in effdb3c2c: when a non-bootstrap source owns an id, bootstrap-owned siblings with that id are dropped from the injected set before policy wrapping, so first-match tool execution can only resolve through the host's source. Mixed-ownership regression test added: host + marked bootstrap sibling ambient with the same id yields exactly one source, listing only the host-credential definition, and the bootstrap factory provably never runs.

**Comment by kwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 4]
```diff
@@ -415,6 +470,27 @@ export function getRuntimeRemoteToolSources(
     }
     return [];
   });
+  // A host-owned id must resolve exclusively to the host's source: tool
+  // execution takes the first source that serves a name, so a bootstrap-owned
+  // sibling sharing the id would bypass the host credential boundary.
+  const hostOwnedInjectedIds = createPrivateSet<string>();
+  for (const source of selectedInjectedSources) {
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts]: Already fixed in e4261b53c (this thread is against the prior head): the host-ownership classification uses own-index iteration instead of for...of, so it no longer dispatches through a caller-controlled Array.prototype[Symbol.iterator]. The remote-source projection intrinsics suite passes (10/10).

File: src/agent/runtime/mcp-server-tool-sources.test.ts [hunk 5]
```diff
@@ -937,3 +938,159 @@ it("keeps injected remote facades out of source collection hooks", async () => {
   assertEquals(await selected[0]!.executeTool("allowed", {}), { ok: true });
   assertEquals(executions, 1);
 });
+
+Deno.test("getRuntimeRemoteToolSources serves a child agent's own named tools past a constrained inherited source", async () => {
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.test.ts]: Fixed in 54acf52a7: the three added cases now register with it() from #veryfront/testing/bdd.ts per AGENTS.md, matching the existing import in this file.

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 6]
```diff
@@ -415,6 +470,27 @@ export function getRuntimeRemoteToolSources(
...
+      selectedInjectedSources,
+      (source) =>
+        !selfServedFirstPartyIds.has(source.id) &&
+        !(isBootstrapIdentityRemoteToolSource(source) && hostOwnedInjectedIds.has(source.id)),
+    ),
+    (source) => {
+      const server = configuredFirstPartyServersBySourceId.get(source.id);
+      const policy = server?.toolPolicy ??
+        (implicitToolNames.length > 0 ? { allow: implicitToolNames } : undefined);
+      return createMcpToolPolicySource(source, policy);
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts]: Fixed in 54acf52a7: the retained-source wrapper now runs through propagateBootstrapIdentity, so an inherited alias this level only policy-narrows keeps its provenance and a deeper delegation can still re-derive its own source for a tool the intermediate policy excluded. Regression test added for the three-level case (root bootstrap alias → intermediate retains and narrows → grandchild selects the alias explicitly): red without the propagation (grandchild resolves no tools), green with it.

**Comment by kwakayama** [comment]
@codex review

**Comment by kwakayama** [comment]
@codex review

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 7]
```diff
@@ -155,7 +188,10 @@ export function constrainRuntimeRemoteToolSources(
 
   const policy = { allow: [...createPrivateSet(allowedToolNames)] };
   const sourcesToConstrain = sources ?? getActiveRuntimeRemoteToolSources() ?? [];
-  return mapPrivateArray(sourcesToConstrain, (source) => createMcpToolPolicySource(source, policy));
+  return mapPrivateArray(
+    sourcesToConstrain,
+    (source) => propagateBootstrapIdentity(source, createMcpToolPolicySource(source, policy)),
+  );
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts:200]: Agreed and fixed in d190029f0. This was a real defect of the change.

Evidence: the `allowedToolNames` argument to `constrainRuntimeRemoteToolSources` is only ever `getRuntimeAllowedRemoteTools(config)`, i.e. `__vfAllowedRemoteTools`. That key is stamped exclusively by trusted run/host code paths (internal run stream from forwarded grants intersected with `runtimeOverrides.toolAllowlist`, the hosted request handler, the hosted chat runtime, fork authorization, and AG-UI retirement limits). It is never a plain agent tool declaration, so every value reaching that wrapper is an authorization ceiling and none of them may be escaped.

Fix: `constrainRuntimeRemoteToolSources` no longer propagates bootstrap provenance onto the policy wrapper. The ceiling wrapper therefore reads as owned at `hostOwnsSourceId`, and a child declaring an excluded first-party tool can no longer rebuild the API source from the process bootstrap identity. Credential-owner binding and the retained-alias wrapper still propagate, which is what this PR actually needs: the local scenario this PR targets stamps no `__vfAllowedRemoteTools` at all, so `constrainRuntimeRemoteToolSources` returns the sources unchanged there.

Regression test `keeps a trusted remote-tool ceiling authoritative for an invoked child`: a bootstrap-built root source serving `list_files` and `create_file` is capped to `list_files`, then a child declaring `create_file` is invoked. Red before the fix (the bootstrap factory ran and the child resolved `[create_file]`), green after (factory never runs, child resolves `[]`). All 34 tests in the file pass, including this PRs four earlier cases.

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 8]
```diff
@@ -415,6 +470,33 @@ export function getRuntimeRemoteToolSources(
     }
     return [];
   });
+  // A host-owned id must resolve exclusively to the host's source: tool
+  // execution takes the first source that serves a name, so a bootstrap-owned
+  // sibling sharing the id would bypass the host credential boundary.
+  const hostOwnedInjectedIds = createPrivateSet<string>();
+  for (let index = 0; index < selectedInjectedSources.length; index++) {
+    const source = selectedInjectedSources[index];
+    if (source !== undefined && !isBootstrapIdentityRemoteToolSource(source)) {
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts:489]: Agreed and fixed in d190029f0. Confirmed the current head avoided iterator dispatch but still dereferenced inherited slots, unlike every helper in `private-array.ts`, which all guard with `Object.hasOwn`.

Fix: the ownership-classification loop now skips non-own indexes before reading the entry, using the module-level captured `Object.hasOwn`.

Regression test `ignores inherited array slots when classifying remote source ownership`: the ambient array has a hole at index 0 with an unmarked same-id source reachable on the prototype, and the real bootstrap-owned source at own index 1. The child declares `list_files` and has no bootstrap identity, so it cannot re-derive, which is the path where the mislabelling is observable. Red before the fix (the bootstrap source was dropped and the child resolved no tools), green after (resolves `[list_files]`). The projection and collection intrinsics suites still pass, 12 tests.

File: src/agent/runtime/mcp-server-tool-sources.ts [hunk 9]
```diff
@@ -376,31 +421,47 @@ export function getRuntimeRemoteToolSources(
...
-        somePrivateArray(
-          injectedSources,
-          (source) => source.id === getFirstPartyMcpSourceId(server),
-        )
-      ) {
+      const sourceId = getFirstPartyMcpSourceId(server);
+      const sameIdSources = filterPrivateArray(
+        injectedSources,
+        (source) => source.id === sourceId,
+      );
```
[the remark this answers is not in the document]
Comment (kwakayama) [reply src/agent/runtime/mcp-server-tool-sources.ts:443]: Good find, and I agree the gap is real, but it is pre-existing on main rather than a defect of this change, so it is filed as veryfront/veryfront-issue-inbox#1339 instead of expanding this PR.

Evidence. I wrote a probe that builds the exact scenario: an ambient `veryfront-platform-mcp` source serving `list_files` and `create_file`, capped to `list_files` through `constrainRuntimeRemoteToolSources`, then a child declaring `tools: { create_file: true }` plus `mcpServers: [{ kind: "veryfront-api", id: "another-id", toolPolicy: { allow: ["create_file"] } }]`. I ran that identical probe in two worktrees.

On the unmodified merge-base b7f552238:
```
PROBE rederivedFromBootstrap = true
PROBE resolved tool names = ["create_file"]
```
On this PR head d190029f0:
```
PROBE rederivedFromBootstrap = true
PROBE resolved tool names = ["create_file"]
```

Identical, so this branch neither introduces nor widens the alias escape. On main the pre-change code already took the same shape: `somePrivateArray(injectedSources, (source) => source.id === getFirstPartyMcpSourceId(server))` finds no match for an aliased id, and the capped source is separately dropped because the selection filter keeps only ambient sources whose id appears in the child configured servers. The probe was deleted, not committed.

What this PR did fix is the exact-id ceiling case, which WAS introduced here: carrying bootstrap provenance onto the cap wrapper. Closing the alias axis needs the ceiling tracked by provenance rather than by id, which is a different mechanism and a larger change than this branch should carry; #1339 records both candidate directions. Resolving on that basis.

**Comment by kwakayama** [comment]
@codex review
